api/sys_raft.go

```diff
@@ -0,0 +1,98 @@
+package api
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"io/ioutil"
+)
+
+type RaftJoinResponse struct {
+	Joined bool `json:"joined"`
+}
+
+type RaftJoinRequest struct {
+	LeaderAddr string `json:"leader_api_addr"`
+	CACert     string `json:"ca_cert":`
+	Retry      bool   `json:"retry"`
+}
+
+func (c *Sys) RaftJoin(opts *RaftJoinRequest) (*RaftJoinResponse, error) {
+	r := c.c.NewRequest("POST", "/v1/sys/storage/raft/join")
+
+	if err := r.SetJSONBody(opts); err != nil {
+		return nil, err
+	}
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+	resp, err := c.c.RawRequestWithContext(ctx, r)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	var result RaftJoinResponse
+	err = resp.DecodeJSON(&result)
+	return &result, err
+}
+
+func (c *Sys) RaftConfig() (*Secret, error) {
+	r := c.c.NewRequest("GET", "/v1/sys/storage/raft/configuration")
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+	resp, err := c.c.RawRequestWithContext(ctx, r)
+
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	return ParseSecret(resp.Body)
+}
+
+func (c *Sys) RaftSnapshot() ([]byte, error) {
+	r := c.c.NewRequest("GET", "/v1/sys/storage/raft/snapshot")
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+	resp, err := c.c.RawRequestWithContext(ctx, r)
+
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	snap, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return snap, nil
+}
+
+func (c *Sys) RaftSnapshotWrite(snapReader io.Reader, force bool) error {
+	snap, err := ioutil.ReadAll(snapReader)
```

Comment:
Again, we don't want to read the snapshot into memory, i think you can just set the request's body = to the reader

Reply:
Fixed.